### PR TITLE
Alerting: Deduplicate receivers during migration

### LINKED ISSUE #36697: Grafana channels are not imported correctly as the contact point
**What happened**:
After enabling `ngalert` on grafana, I see a long long list of random contact points: 
image

**What you expected to happen**:
The actual notification channel is much less and contains other types like `webhook`:
image

**How to reproduce it (as minimally and precisely as possible)**:
enable `nginalert` in `grafana.ini`

**Environment**:
- Grafana version: 8.0.5

### PR DESCRIPTION
**Pull request by codesome**
If 2 panel alerts have the same set of notification channels used, this PR deduplicates them and avoids creating new contact points for such channel groups.

Fixes https://github.com/grafana/grafana/issues/36697

### CODE REVIEW

**Review by kylebrandt — APPROVED** [review]
Looks good to me and works with my instance. Should be easier for users and use to debug with less contact points being created <3

File: pkg/services/sqlstore/migrations/ualert/channel.go [hunk 1]
```diff
@@ -161,6 +177,25 @@ func (m *migration) makeReceiverAndRoute(ruleUid string, channelUids []interface
...
+	uids := make([]string, 0, len(channelUids))
+	for _, u := range channelUids {
+		switch uid := u.(type) {
+		case string:
+			uids = append(uids, uid)
+		case int:
+			uids = append(uids, fmt.Sprintf("%d", uid))
+		default:
+			// Should never happen.
+			return "", errors.New("unknown channel UID type")
```
Comment (papagian) [inline pkg/services/sqlstore/migrations/ualert/channel.go:191]:
```suggestion
		case int, int32, int64:
			uids = append(uids, fmt.Sprintf("%d", uid))
		default:
			// Should never happen.
			return "", fmt.Errorf("unknown channel UID type: %T", u)
```

**Review by papagian — CHANGES_REQUESTED** [review]

**Comment by papagian** [comment]
I've tested it and the duplicates are reduced but I think I found another edge case that can result to duplicates:
if 2 different alerts have only [an obsolete notification channel](https://github.com/grafana/grafana/blob/ac26ca4af7a43f5417aa0e8af7e7f38dffb7e034/pkg/services/sqlstore/migrations/ualert/channel.go#L143-L145) then we will end up 2 contact point with the default channels.

**Comment by codesome** [comment]
@papagian nice catch! Fixed it by filtering the UIDs before getting the key, and also added a log when we are dropping from UID.

**Comment by codesome** [comment]
Since we are deduplicating the contact points, having rule UID in the contact point name would be confusing now and not easy to find. So I have modified the naming of contact point to be `autogen-contact-point-<number>` (also we called it recv before, which can be confusing again, changed now). It should be more friendly to find auto generated contact points now.

Screenshot from 2021-07-29 10-58-00
Screenshot from 2021-07-29 10-58-24

**Comment by papagian** [comment]
It works much better know and I prefer the new naming a lot!
I think we could reduce further the number of policies created for 2 more corner cases (but not completely sure)
Consider the following screenshots:
Screenshot 2021-07-29 at 11 39 39 AM
Screenshot 2021-07-29 at 11 39 32 AM

Rules `UZgwpXW7k`, `XWgQpXZnz`, `jZgQpuW7k` are migrated dashboard alerts that referenced the webhook notifier  (using its `id`) that happened to be also one of the default notification channels.
Rule `DWgwtuZ7z` is migrated dashboard alert that references only an obsolete (non-existing) notifier (using `uid`).

Note also that `autogen-contact-point-2` contains exactly the same contact points with `autogen-contact-point-default`
```
      {
        "name": "autogen-contact-point-default",
        "grafana_managed_receiver_configs": [
          {
            "uid": "TVRwpXW7z",
            "name": "webhook",
            "type": "webhook",
            "disableResolveMessage": false,
            "settings": {
              "autoResolve": true,
              "httpMethod": "POST",
              "password": "",
              "severity": "critical",
              "uploadImage": true,
              "url": "http://localhost:3010",
              "username": ""
            },
            "secureFields": {}
          },
          {
            "uid": "T4RQtXZ7kz",
            "name": "#test-slack-alerts-webhook-url",
            "type": "slack",
            "disableResolveMessage": false,
            "settings": {
              "autoResolve": true,
              "httpMethod": "POST",
              "iconEmoji": "",
              "iconUrl": "",
              "mentionGroups": "",
              "mentionUsers": "",
              "recipient": "#test-alerts",
              "severity": "critical",
              "token": "",
              "uploadImage": true,
              "url": "",
              "username": ""
            },
            "secureFields": {
              "url": true
            }
          }
        ]
      }
```

```
{
        "name": "autogen-contact-point-2",
        "grafana_managed_receiver_configs": [
          {
            "uid": "DZgQtuZnkz",
            "name": "webhook",
            "type": "webhook",
            "disableResolveMessage": false,
            "settings": {
              "autoResolve": true,
              "httpMethod": "POST",
              "password": "",
              "severity": "critical",
              "uploadImage": true,
              "url": "http://localhost:3010",
              "username": ""
            },
            "secureFields": {}
          },
          {
            "uid": "vZgwpuWnzm",
            "name": "#test-slack-alerts-webhook-url",
            "type": "slack",
            "disableResolveMessage": false,
            "settings": {
              "autoResolve": true,
              "httpMethod": "POST",
              "iconEmoji": "",
              "iconUrl": "",
              "mentionGroups": "",
              "mentionUsers": "",
              "recipient": "#test-alerts",
              "severity": "critical",
              "token": "",
              "uploadImage": true,
              "url": "",
              "username": ""
            },
            "secureFields": {
              "url": true
            }
          }
        ]
      },
```

**Review by papagian — APPROVED** [review]
I have tested and it works great!

**Pull request merged**